### ISSUE TITLE
Improve CUE Package Management

### ISSUE THREAD

**Issue by aluzzardi** [dagger/dagger#2163]
Non exhaustive list of requirements, in random order [...]

Question is: should we rewrite our own vendoring system or can we contribute upstream instead?

_Originally posted by @aluzzardi in https://github.com/dagger/dagger/discussions/2052#discussioncomment-2527660_

---

## Essential

These are focused on freeing us from embedding the built-in packages in the binary, and from needing to vendor packages in `cue.mod/pkg` (i.e., committing them to VCS).

- [x] #2413
- [ ] #2631
- [ ] #2672
- [x] #2502
- [x] #2504
- [x] #2506
- [x] #2503
- [x] ~#2431~

## Enhancements

Incremental improvements mostly from the upstream proposal, until we have an official package management solution.

- [x] #2496
- [x] #2508
- [x] #2511
  - [x] #2034
- [x] #2509
- [x] #2510

## Upstream

- https://github.com/dagger/dagger/discussions/857
- https://github.com/cue-lang/cue/issues/851

## Docs
- [x] #2186
- [x] #2315
- [x] #2543

**Comment by jlongtine** [comment]
I'd definitely vote for trying to contribute upstream, if we can.

**Comment by helderco** [comment]
Ref https://github.com/dagger/dagger/issues/2186#issuecomment-1104570150

**Comment by samalba** [comment]
@morlay shared his project on Discord: https://github.com/octohelm/cuemod - maybe we could use it to improve our package support?

@morlay is cuemod embeddable into other projects using CUE?

**Comment by morlay** [comment]
@samalba Yes it is.

I could just use with  `github.com/octohelm/cuemod/pkg/cuemod.ContextFor` https://github.com/octohelm/cuemod/blob/main/pkg/cuemod/context_test.go#L25

But the spec is different from `dagger.mod` https://github.com/octohelm/cuemod#spec-cuemodmodulecue

**Comment by morlay** [comment]
In fact. the two tools could be split to use.

https://github.com/morlay/dagger-cuemod-example


once `dagger.io/*` and `universe.dagger.io/*` support `go-get`, We could remove the replace.

```cue
replace: {
	"dagger.io":          "github.com/dagger/dagger/pkg/dagger.io"
	"universe.dagger.io": "github.com/dagger/dagger/pkg/universe.dagger.io"
}
```

**Comment by morlay** [comment]
@samalba 

With some refactor. now just change oneline in `compiler/build.go` (github.com/octohelm/cuemod@v0.3.4)

```go
buildConfig := cuemod.ContextFor(src).BuildConfig(context.Background())
```

Before the "go-get" supported https://github.com/dagger/dagger/issues/2163#issuecomment-1105189593
the replace directive have to added.

**Comment by aluzzardi** [comment]
That's super interesting @morlay, we should definitely look into it /cc @helderco 

Do you think we could use `cue-get=1` / `cue-import` / `cue-source` instead of `go-*`? I believe that's how CUE is going to implement it https://github.com/cue-lang/cue/issues/851 (`https://acme.com/quote?cue-get=1` from the spec) /cc @jlongtine

**Comment by jlongtine** [comment]
That is my understanding of how `cue mod` will work, but I don't think any system will respect it yet, including GitHub. :/

**Comment by morlay** [comment]
@aluzzardi 

i think we could support both `go-get` and `cue-get`.

just need put those meta in html.
query parameters could be ignored.

Now we could use the channel from go,
private vcs like gitlab already support `go-get`.

this will be easy to start.

**Comment by helderco** [comment]
Thanks @morlay, I just need to finish something else before properly evaluating this.

**Comment by aluzzardi** [comment]
Hey @morlay, haven't had a chance to evaluate this yet, but had a few questions in the meantime:

- Are the actual modules stored in `cue.mod/pkg` or are they somewhere else (if so, are you then using an overlay)?

- Is `module.sum` used to enforce the correctness of packages (e.g. making sure they have not been modified), or are we just storing the checksum for now? (in our old `mod` code, we were computing checksums to verify everything was correct before loading CUE: https://github.com/dagger/dagger/blob/main/mod/file.go#L255)

- > as @jlongtine said, no vcs support cue-get now.
  - This I'm not sure to understand: why would GitHub need to support `?go-get=1`?
  - I thought `?go-get=1` was ONLY used for vanity URLs
  - VCS support is hardcoded inside Go, there's URL translators for all popular VCS: https://github.com/golang/go/blob/master/src/cmd/go/internal/vcs/vcs.go#L1437
  - Does GitHub do something special with `go-get=1`?

-  Runtime version
   - Could we store a "dagger" field in `module.cue`? Or should we use a different file
   -  See https://github.com/dagger/dagger/pull/2321#issuecomment-1111578974 for context
   - Basically, similar to `go` which sets a field like `go 1.16` in `go.mod`, we'd like to add a field to `module.cue` (or similar) that we could use for compatibility checks

**Comment by morlay** [comment]
Hi @aluzzardi 

> * Are the actual modules stored in `cue.mod/pkg` or are they somewhere else (if so, are you then using an overlay)?

Code store under `GOMODCACHE`, here a real world project with `dagger` and `k8s`. 

* `gen` contains the genereted cue files from other languages.
* `pkg` just create symlinks from `GOMODCACHE`, the behavior learned from [pnpm](https://github.com/pnpm/pnpm).

The both  be git-ignored.
```
./cue.mod
├── gen
│   └── k8s.io
│       ├── api
│       │   ├── apps
│       │   │   └── v1
│       │   │       ├── register_go_gen.cue
│       │   │       └── types_go_gen.cue
│       │   ├── core
│       │   │   └── v1
│       │   │       ├── annotation_key_constants_go_gen.cue
│       │   │       ├── register_go_gen.cue
│       │   │       ├── types_go_gen.cue
│       │   │       ├── well_known_labels_go_gen.cue
│       │   │       └── well_known_taints_go_gen.cue
│       │   ├── networking
│       │   │   └── v1
│       │   │       ├── register_go_gen.cue
│       │   │       ├── types_go_gen.cue
│       │   │       └── well_known_annotations_go_gen.cue
│       │   └── rbac
│       │       └── v1
│       │           ├── register_go_gen.cue
│       │           └── types_go_gen.cue
│       └── apimachinery
│           └── pkg
│               ├── api
│               │   └── resource
│               │       ├── amount_go_gen.cue
│               │       └── quantity_go_gen.cue
│               ├── apis
│               │   └── meta
│               │       └── v1
│               │           ├── duration_go_gen.cue
│               │           ├── group_version_go_gen.cue
│               │           ├── meta_go_gen.cue
│               │           ├── micro_time_go_gen.cue
│               │           ├── register_go_gen.cue
│               │           ├── time_go_gen.cue
│               │           ├── time_proto_go_gen.cue
│               │           ├── types_go_gen.cue
│               │           └── watch_go_gen.cue
│               ├── runtime
│               │   ├── codec_go_gen.cue
│               │   ├── converter_go_gen.cue
│               │   ├── helper_go_gen.cue
│               │   ├── interfaces_go_gen.cue
│               │   ├── negotiate_go_gen.cue
│               │   ├── swagger_doc_generator_go_gen.cue
│               │   ├── types_go_gen.cue
│               │   └── types_proto_go_gen.cue
│               ├── types
│               │   ├── namespacedname_go_gen.cue
│               │   ├── nodename_go_gen.cue
│               │   ├── patch_go_gen.cue
│               │   └── uid_go_gen.cue
│               ├── util
│               │   └── intstr
│               │       └── intstr_go_gen.cue
│               └── watch
│                   ├── filter_go_gen.cue
│                   ├── mux_go_gen.cue
│                   ├── streamwatcher_go_gen.cue
│                   └── watch_go_gen.cue
├── module.cue
├── module.sum
└── pkg
    ├── dagger.io
    │   └── dagger
    │       ├── compat.cue -> /Users/morlay/pkg/mod/github.com/dagger/dagger@v0.2.7/pkg/dagger.io/dagger/compat.cue
    │       ├── core -> /Users/morlay/pkg/mod/github.com/dagger/dagger@v0.2.7/pkg/dagger.io/dagger/core
    │       ├── plan.cue -> /Users/morlay/pkg/mod/github.com/dagger/dagger@v0.2.7/pkg/dagger.io/dagger/plan.cue
    │       ├── types.cue -> /Users/morlay/pkg/mod/github.com/dagger/dagger@v0.2.7/pkg/dagger.io/dagger/types.cue
    │       └── values.cue -> /Users/morlay/pkg/mod/github.com/dagger/dagger@v0.2.7/pkg/dagger.io/dagger/values.cue
    └── universe.dagger.io
        └── docker
            ├── build.cue -> /Users/morlay/pkg/mod/github.com/dagger/dagger@v0.2.7/pkg/universe.dagger.io/docker/build.cue
            ├── cli
            │   ├── client.cue -> /Users/morlay/pkg/mod/github.com/dagger/dagger@v0.2.7/pkg/universe.dagger.io/docker/cli/client.cue
            │   ├── load.cue -> /Users/morlay/pkg/mod/github.com/dagger/dagger@v0.2.7/pkg/universe.dagger.io/docker/cli/load.cue
            │   └── test -> /Users/morlay/pkg/mod/github.com/dagger/dagger@v0.2.7/pkg/universe.dagger.io/docker/cli/test
            ├── image.cue -> /Users/morlay/pkg/mod/github.com/dagger/dagger@v0.2.7/pkg/universe.dagger.io/docker/image.cue
            ├── pull.cue -> /Users/morlay/pkg/mod/github.com/dagger/dagger@v0.2.7/pkg/universe.dagger.io/docker/pull.cue
            ├── push.cue -> /Users/morlay/pkg/mod/github.com/dagger/dagger@v0.2.7/pkg/universe.dagger.io/docker/push.cue
            ├── run.cue -> /Users/morlay/pkg/mod/github.com/dagger/dagger@v0.2.7/pkg/universe.dagger.io/docker/run.cue
            ├── set.cue -> /Users/morlay/pkg/mod/github.com/dagger/dagger@v0.2.7/pkg/universe.dagger.io/docker/set.cue
            └── test -> /Users/morlay/pkg/mod/github.com/dagger/dagger@v0.2.7/pkg/universe.dagger.io/docker/test
```


> * Is `module.sum` used to enforce the correctness of packages (e.g. making sure they have not been modified), or are we just storing the checksum for now? (in our old `mod` code, we were computing checksums to verify everything was correct before loading CUE: https://github.com/dagger/dagger/blob/main/mod/file.go#L255)

Yes. `module.sum` is same as `go.sum`, it should be 
But sames i miss to check it with stored importPath sums (I need fix it) https://github.com/octohelm/cuemod/blob/main/pkg/cuemod/mod_resolver.go#L349 

>   * VCS support is hardcoded inside Go, there's URL translators for all popular VCS: https://github.com/golang/go/blob/master/src/cmd/go/internal/vcs/vcs.go#L1437
>   * Does GitHub do something special with `go-get=1`?

As popular vcs, Github not need to do anything. 
But for private vcs with custom domain, here need be the special meta to tell golang this `importPath` could be clone from.

Example  from my gitlab self-hosted
```
curl https://git.innoai.tech/infrav2/hx\?go-get\=1
<html><head><meta name="go-import" content="git.innoai.tech/infrav2/hx git https://git.innoai.tech/infrav2/hx.git" /><meta name="go-source" content="git.innoai.tech/infrav2/hx https://git.innoai.tech/infrav2/hx https://git.innoai.tech/infrav2/hx/-/tree/master{/dir} https://git.innoai.tech/infrav2/hx/-/blob/master{/dir}/{file}#L{line}" /></head><body>go get https://git.innoai.tech/infrav2/hx</body></html>% 
```

>   * Could we store a "dagger" field in `module.cue`? Or should we use a different file

Not need this.
Once mod landed in dagger.  the version will be in `require["dagger.io/dagger"]`

https://github.com/octohelm/cuemod/blob/dagger/cue.mod/module.cue
	
```
module: "github.com/octohelm/cuemod"

require: {
	"github.com/dagger/dagger": "v0.2.8"
	"dagger.io/dagger":         "v0.2.8"  // <---- here should be added once use dagger
	"k8s.io/api":               "v0.23.5"
	"k8s.io/apimachinery":      "v0.23.5"
}

replace: {
	"dagger.io":          "github.com/dagger/dagger/pkg/dagger.io"
	"universe.dagger.io": "github.com/dagger/dagger/pkg/universe.dagger.io"
}

replace: {
	"k8s.io/api":          "" @import("go")
	"k8s.io/apimachinery": "" @import("go")
}
```

**Comment by aluzzardi** [comment]
Thank you for all the info!

- Understood for private gitlab and the need for go-get

- A downside is it might clash with Go vanity URLs, but maybe that's okay (with dagger.io for instance, Go needs dagger/dagger whereas for CUE it's dagger/dagger/pkg/dagger.io)

- Curious as to why we need to use GOMODCACHE instead of another directory (for example $HOME/.cache/cue)? Won't it clash with go?

- Using the symlinks is clever, and makes cue eval continue working. I'm curious why though every file is individually symlinked instead of the whole directory?

- Also I see the symlink points to dagger@v0.2.7. Shouldn't that be the sum instead of the tag? Not sure how Go itself handles this 

- dagger field: I was thinking of how a module can specify runtime version requirement (similar to "go 1.16" at the beginning of a go.mod file). But perhaps you're right, this can be enough

**Comment by morlay** [comment]
> * A downside is it might clash with Go vanity URLs, but maybe that's okay (with dagger.io for instance, Go needs dagger/dagger whereas for CUE it's dagger/dagger/pkg/dagger.io)

it just tell go to where to download the whole repo code. and the meta will location currect sub path (notice the `go-source`)

```
<meta name="go-import" content="dagger.io/dagger git https://github.com/dagger/dagger">
<meta name="go-source" content="dagger.io/dagger https://github.com/dagger/dagger https://github.com/dagger/dagger/tree/main/pkg/dagger.io/dagger{/dir} https://github.com/dagger/dagger/blob/main/pkg/dagger.io/dagger{/dir}/{file}#L{line}">
```

> * Curious as to why we need to use GOMODCACHE instead of another directory (for example $HOME/.cache/cue)? 

Just because I use go mod's code directly for quick implements, once cue team support cue mod, it will be `CUEMODCACHE`.

> Won't it clash with go?

Nope, it works well.

> * Using the symlinks is clever, and makes cue eval continue working. I'm curious why though every file is individually symlinked instead of the whole directory?

To avoid symlink creating conflicts. Example:

```
dagger.io/  
    dagger/  <-- if here create symlink
        core/  <-- When process on the sub mod, create symlink here will create self-symlink, because current folder already in code which store
```

And I don't want to check each folder symlink creatable. so just follow pnpm did, which already works well in node.

> * Also I see the symlink points to dagger@v0.2.7. Shouldn't that be the sum instead of the tag? Not sure how Go itself handles this

Nope. `sum` still be used. 

go mod will check the downloaded package by GOSUMDB first (https://go.dev/ref/mod#checksum-database)
to make sure download is safe
Then extract it into `GOMODCACHE` with version in-path, and quick calc the hash (https://go.dev/ref/mod#go-sum-files).
to make sure the extracted codes is correct.

I think it is safe enough, then i just creating symlinks from `GOMODCACHE`

**Comment by morlay** [comment]
@aluzzardi @helderco 
I have removed most magic things, `github.com/octohelm/cuemod v0.4.0`
only focus about cue mod.

and only remain the special feature, generating cue codes to from golang types.
```
replace: {
	"k8s.io/api":          "" @import("go")
	"k8s.io/apimachinery": "" @import("go")
}
```

**Comment by helderco** [comment]
Thanks @morlay, I'm going to give it a test drive.

**Comment by shykes** [comment]
Added #2496

**Comment by helderco** [comment]
I've converted a few list items into issues.

**Comment by mircubed** [comment]
@helderco to clean up for the new year, can you determine what needs to be kept for this large issue or maybe we can just close it?

**Comment by helderco** [comment]
This should only be relevant if we support any version using CUE. However, with the focus on the GraphQL based SDKs, the CUE package manager is no longer a major onboarding issue when adopting Dagger.

Every other SDK uses their language's package manager. CUE is the exception because there isn't one yet, but there are some third party attempts for one and I don't think it's worth it for us to be another. The ideal situation has always been that the CUE project releases an official solution for its community.

The current CUE SDK v0.2 is frozen and I'm not sure when we'll release a new one, but let's reassess the upstream situation then. In the meantime, you can still try some other tool, vendor manually or use the current implementation to the extent that it works for you.

I'm going to close some of the issues in this epic but since there's still quite a few CUE related issues that are open I'm going to also leave open some of the more essential fixes and hand this over to @jlongtine to decide what to do with what's left.

**Comment by shykes** [comment]
Agreed completely.

**Comment by TomChv** [comment]
I'm closing this issue since we no longer support the cue SDK
